File: backend/src/test/java/lan/dk/podcastserver/entity/ItemTest.java
package lan.dk.podcastserver.entity;

import org.junit.Before;
import org.junit.Test;
import org.springframework.util.FileSystemUtils;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.ZonedDateTime;
import java.util.UUID;

import static lan.dk.podcastserver.assertion.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThat;

/**
 * Created by kevin on 15/06/15 for HackerRank problem
 */
public class ItemTest {

    private static final UUID PODCAST_ID = UUID.randomUUID();
    private static Item ITEM = new Item();
    private static Podcast PODCAST = new Podcast().setId(UUID.randomUUID());
    private static final ZonedDateTime NOW = ZonedDateTime.now();
    private static final Cover COVER = Cover.builder().url("http://fakeItem.com/cover.png").build();
    private static final Cover PODCAST_COVER = Cover.builder().url("PodcastCover").build();
    private static final UUID ID = UUID.randomUUID();

    @Before
    public void beforeEach() throws IOException {
        /* Given */
        PODCAST = Podcast.builder()
                .title("Fake Podcast")
                .id(PODCAST_ID)
                .type("Youtube")
                .cover(PODCAST_COVER)
                .hasToBeDeleted(Boolean.TRUE)
                .build();


        ITEM = new Item()
                .setId(ID)
                .setTitle("Fake Item")
                .setUrl("http://fakeItem.com")
                .setPodcast(PODCAST)
                .setPubDate(NOW)
                .setDescription("Fake item description")
                .setMimeType("video/mp4")
                .setLength(123456L)
                .setCover(COVER)
                .setFileName("fakeItem.mp4");

        Item.rootFolder = Paths.get("/tmp/podcast");

        FileSystemUtils.deleteRecursively(Item.rootFolder.toFile());
        Files.createDirectories(ITEM.getLocalPath().getParent());
        Files.createFile(ITEM.getLocalPath());
    }

    @Test
    public void should_have_be_initialazed() {
        /* When */
        /* Then */
        assertThat(ITEM)
                .hasId(ID)
                .hasTitle("Fake Item")
                .hasUrl("http://fakeItem.com")
                .hasPodcast(PODCAST)
                .hasPubDate(NOW)
                .hasDescription("Fake item description")
                .hasMimeType("video/mp4")
                .hasLength(123456L)
                .hasCover(COVER)
                .hasFileName("fakeItem.mp4")
                .hasStatus(Status.NOT_DOWNLOADED)
                .hasProgression(0);
    }

    @Test
    public void should_change_his_status() {

        /* When */ ITEM.setStatus(Status.PAUSED);
        /* Then */ assertThat(ITEM)
                .hasStatus(Status.PAUSED);

        /* When */ ITEM.setStatus(Status.FINISH);
        /* Then */ assertThat(ITEM)
                .hasStatus(Status.FINISH);
    }

    @Test
    public void should_advance_in_progression() {
        /* When */ ITEM.setProgression(50);
        /* Then */ assertThat(ITEM).hasProgression(50);
    }

    @Test
    public void should_set_the_downloaddate() {
        /* Given */ ZonedDateTime downloaddate = ZonedDateTime.now();
        /* When */  ITEM.setDownloadDate(downloaddate);
        /* Then */  assertThat(ITEM).hasDownloadDate(downloaddate);
    }

    @Test
    public void should_increment_the_number_of_retry() {
        assertThat(ITEM).hasNumberOfFail(0);

        ITEM.setNumberOfFail(6);
        assertThat(ITEM).hasNumberOfFail(6);

        ITEM.addATry();
        assertThat(ITEM).hasNumberOfFail(7);
    }

    @Test
    public void should_have_a_valid_url() {
        assertThat(ITEM.hasValidURL()).isTrue();

        PODCAST.setType("upload");
        ITEM.setUrl("");
        assertThat(ITEM.hasValidURL()).isTrue();

        PODCAST.setType("Youtube");
        ITEM.setUrl("");
        assertThat(ITEM.hasValidURL()).isFalse();

    }

    @Test
    public void should_report_parent_podcast_id() {
        assertThat(ITEM.getPodcastId()).isEqualTo(PODCAST.getId());

        ITEM.setPodcast(null);
        assertThat(ITEM.getPodcastId())
                .isNotEqualTo(PODCAST.getId())
                .isNull();
    }

    @Test
    public void should_report_parent_podcast_cover() {
        assertThat(ITEM.getCoverOfItemOrPodcast())
                .isEqualTo(Cover.builder().url("/api/podcasts/" + PODCAST_ID + "/items/" + ID + "/cover.png").build());

        ITEM.setCover(null);
        assertThat(ITEM.getCoverOfItemOrPodcast()).isSameAs(PODCAST_COVER);
    }

    @Test
    public void should_expose_the_API_url() {
        assertThat(ITEM.getProxyURLWithoutExtention())
                .isEqualTo(String.format("/api/podcasts/%s/items/%s/" + "Fake_Item", PODCAST_ID, ID));

        assertThat(ITEM.getProxyURL())
                .isEqualTo(String.format("/api/podcasts/%s/items/%s/Fake_Item.mp4", PODCAST_ID, ID));
    }

    @Test
    public void should_get_the_local_path() {
        /* When */ Path localPath = ITEM.getLocalPath();
        /* Then */ assertThat(localPath).hasFileName("fakeItem.mp4");

        /* When */ String localStringPath = ITEM.getLocalUri();
        /* Then */ assertThat(localStringPath).isEqualTo("/tmp/podcast/Fake Podcast/fakeItem.mp4");
    }

    @Test
    public void should_expose_isDownloaded() {
        assertThat(ITEM.isDownloaded()).isTrue();

        ITEM.setFileName(null);
        assertThat(ITEM.isDownloaded()).isFalse();
    }

    @Test
    public void should_be_reset() {
        /* Given */ ITEM.setStatus(Status.FINISH);
        /* When  */ ITEM.reset();
        /* Then  */ assertThat(ITEM).hasFileName(null)
                .hasDownloadDate(null)
                .hasStatus(Status.NOT_DOWNLOADED);
    }

    @Test
    public void should_change_the_local_uri() {
        /* When */ ITEM.setLocalUri("http://www.google.fr/mavideo.mp4");
        /* Then */ assertThat(ITEM).hasFileName("mavideo.mp4");
    }

    @Test
    public void should_equals_and_hashcode() {
        /* Given */
        Object anObject = new Object();
        Item withSameId = new Item().setId(ID);
        Item withSameUrl = new Item().setUrl(ITEM.getUrl());
        Item withSameName = new Item().setUrl("http://test.domain.com/toto/fakeItem.com");
        /*Item withSameLocalUri = new Item().setPodcast(PODCAST).setFileName("fakeItem.mp4");*/

        /* Then */
        assertThat(ITEM)
            .isEqualTo(ITEM)
            .isNotEqualTo(anObject)
            .isEqualTo(withSameId)
            .isEqualTo(withSameUrl)
            .isNotEqualTo(withSameName)
            /*.isEqualTo(withSameLocalUri)*/;

        assertThat(ITEM.hashCode()).isEqualTo(new Item().setUrl(ITEM.getUrl()).setPubDate(NOW).hashCode());

    }

    @Test
    public void should_toString() {
        assertThat(ITEM.toString())
                .isEqualTo("Item{id="+ ID +", title='Fake Item', url='http://fakeItem.com', pubDate="+ NOW +", description='Fake item description', mimeType='video/mp4', length=123456, status='NOT_DOWNLOADED', progression=0, downloaddate=null, podcast=Podcast{id="+PODCAST_ID+", title='Fake Podcast', url='null', signature='null', type='Youtube', lastUpdate=null}, numberOfTry=0}");

    }

    @Test
    public void should_delete() {
        /* Given */ ITEM.setStatus(Status.FINISH);
        /* When  */ ITEM.deleteDownloadedFile();
        /* Then  */ assertThat(ITEM).hasFileName(null)
                .hasStatus(Status.DELETED);
    }

    @Test
    public void should_preremove() throws IOException {
        /* Given */
        Path fileToDelete = ITEM.getLocalPath();

        /* When */
        ITEM.preRemove();

        /* Then */
        assertThat(fileToDelete).doesNotExist();
    }

    @Test
    public void should_generate_exception_on_deletion() {
        /* Given */
        Item.rootFolder = Paths.get("/");
        PODCAST.setTitle("sbin");
        ITEM.setFileName("fsck");

        /* When */  ITEM.deleteDownloadedFile();
    }

}

File: backend/src/test/kotlin/com/github/davinkevin/podcastserver/entity/ItemTest.kt
package com.github.davinkevin.podcastserver.entity

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Disabled
import org.junit.jupiter.api.Test
import org.springframework.util.FileSystemUtils
import java.io.IOException
import java.lang.Boolean.TRUE
import java.nio.file.Files
import java.nio.file.Paths
import java.time.ZonedDateTime
import java.util.*

/**
 * Created by kevin on 15/06/15 for HackerRank problem
 */
class ItemTest {

    @BeforeEach
    fun beforeEach() {
        /* Given */
        PODCAST = Podcast()
        PODCAST.title = "Fake Podcast"
        PODCAST.id = PODCAST_ID
        PODCAST.type = "Youtube"
        PODCAST.cover = PODCAST_COVER
        PODCAST.hasToBeDeleted = TRUE


        ITEM = Item()
        ITEM.id = ID
        ITEM.title = "Fake Item"
        ITEM.url = "http://fakeItem.com"
        ITEM.podcast = PODCAST
        ITEM.pubDate = NOW
        ITEM.description = "Fake item description"
        ITEM.mimeType = "video/mp4"
        ITEM.length = 123456L
        ITEM.cover = COVER
        ITEM.fileName = "fakeItem.mp4"

        Item.rootFolder = Paths.get("/tmp/podcast")

        FileSystemUtils.deleteRecursively(Item.rootFolder!!.toFile())
        Files.createDirectories(ITEM.localPath.parent)
        Files.createFile(ITEM.localPath)
    }

    @Test
    fun `should have be initialazed`() {
        /* When */
        /* Then */
        assertThat(ITEM.id).isEqualTo(ID)
        assertThat(ITEM.title).isEqualTo("Fake Item")
        assertThat(ITEM.url).isEqualTo("http://fakeItem.com")
        assertThat(ITEM.podcast).isEqualTo(PODCAST)
        assertThat(ITEM.pubDate).isEqualTo(NOW)
        assertThat(ITEM.description).isEqualTo("Fake item description")
        assertThat(ITEM.mimeType).isEqualTo("video/mp4")
        assertThat(ITEM.length).isEqualTo(123456L)
        assertThat(ITEM.cover).isEqualTo(COVER)
        assertThat(ITEM.fileName).isEqualTo("fakeItem.mp4")
        assertThat(ITEM.status).isEqualTo(Status.NOT_DOWNLOADED)
        assertThat(ITEM.progression).isEqualTo(0)
    }

    @Test
    fun `should change his status`() {

        /* When */ ITEM.status = Status.PAUSED
        /* Then */ assertThat(ITEM.status).isEqualTo(Status.PAUSED)

        /* When */ ITEM.status = Status.FINISH
        /* Then */ assertThat(ITEM.status).isEqualTo(Status.FINISH)
    }

    @Test
    fun `should advance in progression`() {
        /* When */ ITEM.progression = 50
        /* Then */ assertThat(ITEM.progression!!).isEqualTo(50)
    }

    @Test
    fun `should set the downloaddate`() {
        /* Given */
        val downloaddate = ZonedDateTime.now()
        /* When */  ITEM.downloadDate = downloaddate
        /* Then */  assertThat(ITEM.downloadDate).isEqualTo(downloaddate)
    }

    @Test
    fun `should_increment_the number of retry`() {
        assertThat(ITEM.numberOfFail).isEqualTo(0)

        ITEM.numberOfFail = 6
        assertThat(ITEM.numberOfFail!!).isEqualTo(6)

        ITEM.addATry()
        assertThat(ITEM.numberOfFail!!).isEqualTo(7)
    }

    @Test
    fun `should_have a valid url`() {
        assertThat(ITEM.hasValidURL()).isTrue()

        PODCAST.type = "upload"
        ITEM.url = ""
        assertThat(ITEM.hasValidURL()).isTrue()

        PODCAST.type = "Youtube"
        ITEM.url = ""
        assertThat(ITEM.hasValidURL()).isFalse()

    }

    @Test
    fun `should_report parent podcast id`() {
        assertThat(ITEM.podcastId).isEqualTo(PODCAST.id)

        ITEM.podcast = null
        assertThat(ITEM.podcastId)
                .isNotEqualTo(PODCAST.id)
                .isNull()
    }

    @Test
    @Disabled("will be fully removed when Item entity will be removed")
    fun `should_report parent podcast cover`() {
        val cover = Cover()
        cover.url = "/api/podcasts/$PODCAST_ID/items/$ID/cover.png"

        assertThat(ITEM.coverOfItemOrPodcast)
                .isEqualTo(cover)

        ITEM.cover = null
        assertThat(ITEM.coverOfItemOrPodcast).isSameAs(PODCAST_COVER)
    }

    @Test
    fun `should_expose the API url`() {
        assertThat(ITEM.proxyURLWithoutExtention)
                .isEqualTo(String.format("/api/v1/podcasts/%s/items/%s/" + "Fake_Item", PODCAST_ID, ID))

        assertThat(ITEM.proxyURL)
                .isEqualTo(String.format("/api/v1/podcasts/%s/items/%s/Fake_Item.mp4", PODCAST_ID, ID))
    }

    @Test
    fun `should_get the local path`() {
        /* When */
        val localPath = ITEM.localPath
        /* Then */ assertThat(localPath).hasFileName("fakeItem.mp4")

        /* When */
        val localStringPath = ITEM.localUri
        /* Then */ assertThat(localStringPath).isEqualTo("/tmp/podcast/Fake Podcast/fakeItem.mp4")
    }

    @Test
    fun `should expose isDownloaded`() {
        assertThat(ITEM.isDownloaded).isTrue()

        ITEM.fileName = null
        assertThat(ITEM.isDownloaded).isFalse()
    }

    @Test
    fun `should_change the local uri`() {
        /* When */
        ITEM.localUri = "http://www.google.fr/mavideo.mp4"
        /* Then */
        assertThat(ITEM.fileName).isEqualTo("mavideo.mp4")
    }

    @Test
    fun `should equals and hashcode`() {
        /* Given */
        val anObject = Any()
        val withSameId = Item()
        withSameId.id = ID
        val withSameUrl = Item()
        withSameUrl.url = ITEM.url
        val withSameName = Item()
        withSameName.url = "http://test.domain.com/toto/fakeItem.com"
        /*Item withSameLocalUri = new Item().setPodcast(PODCAST).setFileName("fakeItem.mp4");*/

        /* Then */
        assertThat(ITEM)
                .isEqualTo(ITEM)
                .isNotEqualTo(anObject)
                .isEqualTo(withSameId)
                .isEqualTo(withSameUrl)
                .isNotEqualTo(withSameName)/*.isEqualTo(withSameLocalUri)*/

        val expected = Item()
        expected.url = ITEM.url
        expected.pubDate = NOW
        assertThat(ITEM.hashCode()).isEqualTo(expected.hashCode())

    }

    @Test
    fun `should toString`() {
        assertThat(ITEM.toString())
                .isEqualTo("Item{id=$ID, title='Fake Item', url='http://fakeItem.com', pubDate=$NOW, description='Fake item description', mimeType='video/mp4', length=123456, status='NOT_DOWNLOADED', progression=0, downloaddate=null, podcast=Podcast{id=$PODCAST_ID, title='Fake Podcast', url='null', signature='null', type='Youtube', lastUpdate=null}, numberOfTry=0}")

    }

    @Test
    fun `should delete`() {
        /* Given */
        ITEM.status = Status.FINISH
        /* When  */
        ITEM.deleteDownloadedFile()
        /* Then  */
        assertThat(ITEM.fileName).isEqualTo(null)
        assertThat(ITEM.status).isEqualTo(Status.DELETED)
    }

    @Test
    @Throws(IOException::class)
    fun `should preremove`() {
        /* Given */
        val fileToDelete = ITEM.localPath

        /* When */
        ITEM.preRemove()

        /* Then */
        assertThat(fileToDelete).doesNotExist()
    }

    @Test
    fun `should_generate exception on deletion`() {
        /* Given */
        Item.rootFolder = Paths.get("/")
        PODCAST.title = "sbin"
        ITEM.fileName = "fsck"

        /* When */  ITEM.deleteDownloadedFile()
    }

    companion object {

        private val PODCAST_ID = UUID.randomUUID()
        private var ITEM = Item()
        private var PODCAST: Podcast = object : Podcast() {
            init {
                id = UUID.randomUUID()
            }
        }
        private val NOW = ZonedDateTime.now()
        private val COVER = object : Cover() {
            init {
                url = "http://fakeItem.com/cover.png"
            }
        }
        private val PODCAST_COVER = object : Cover() {
            init {
                url = "PodcastCover"
            }
        }
        private val ID = UUID.randomUUID()
    }

}
